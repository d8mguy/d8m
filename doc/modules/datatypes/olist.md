## Olist

Module olist provides ordered lists, i.e., lists stored in ascending order on `$<`. These are fairly different from regular lists, and the methods are different from those of list. In fact, olists as done here are really implementing sets, and the module includes an STMap statement so this can happen automatically. The correctness of this depends on assumptions discussed in the next few paragraphs.

There are two typegen definitions here: of `olist` and `olist0`. `Olist0` is for ordered types (types with a `$<` method); it
has a different identifier because typegens don't overload in d8m. The `olist` typegen has 3 arguments: the element type `T`, an ordered type that elements can be extracted to, and a key extraction function `oextract`. Thus, for `olist`, the type `T` is not ordered but `oextract` finds something within it that is. In more standard database language, oextract provides the _key_. 

Now let's talk about assumptions made. First, in d8m we always assume that the `$<` symbol designates a total order (i.e., not a partial order). This is defined to mean that every pair of elements is comparable, that is, for each pair of elements `t1`, `t2` we have `t1 < t2 || t1 == t2 || t2 < t1`. 

Second, in `olist` specifically, we assume that `oextract` is injective when restricted to the domain of any given olist. A function `f` is injective just in case for any `t1`, `t2`, `f(t1)` `==` `f(t2)` implies `t1` `==` `t2`.  

In practice, `olist0` will be used for totally ordered base types like numbers and strings, plus tuple types provided with a `$<` method, while `olist` will be used for tuple types where we don't want to fix the ordering into the type itself, but the tuple type contains one or more ordered attributes that can serve as keys. Consider `accts:list(T)` where `T` has various attributes, including 

    attribute email:string
Now imagine defining `oextract` as

    \(x:T) { x.email }
In database language, we're saying the `email` attribute is the key. And in that mindset, we know that when adding items to `accts`, we need to ensure that the `email` attribute is never duplicated. 

This is precisely what it means in mathematical language to say that `oextract` is injective, when restricted to `accts`. As a tuple type, it's easy to create two distinct items with the same `email` attribute. That means that as a mathematical function, `oextract` is not injective. It's in the nature of tuple types that it cannot be. So the "restricted to `accts`" clause is important here. It means that although the tuple type `T` can easily have distinct elements with the same value of `email`, the collection of `T`'s in `accts` will not. 

It should be easy to see why this condition isn't expressible within d8m's type language (and arguably, isn't a proper thing for type systems to attempt to express at all). D8m's type system is about the structure of entities; this condition concerns constraints on their contents. 

Making this assumption ensures that there cannot be consecutive elements in the olist that map to the same key but are distinct, which assures us that if we're looking for some `x:T` in `accts` and find one whose `oextract` matches at index `k`, then `x` is in `accts` iff `accts[k] == x`. If in addition, we insert into the list only elements that aren't already in it, then we know the list is a set. `Olist` provides an `insertIfUniq` method to make this easy, and generally assumes you don't insert duplicates. However, it doesn't enforce this (for example, by checking for duplication in the `insert` method) for two reasons. First, it's quite common in practice to have code structured so that duplicate insertion is impossible; in this case, embedding a check in `insert` loses performance for no reason. Second, if a duplication check in `insert` fails, the correct response depends on the context &mdash; it might be to panic, to return silently, or return some kind of error message. The solution I've adopted is to explain all this in detail and then make both methods available: `insert` for when the code structure makes duplication impossible; `insertIfUniq` for when a duplicate should be ignored. 

An alternative design choice would be to drop the "injective when restricted" condition. Doing this would weaken one of the really attractive things about olists, which is that the worst case lookup time is `O(lg(n))`. Since you're not guaranteed that the element returned by binary search is the element you're looking for, you must follow each binary search with a linear one that continues until the key changes in the olist. In general, this linear search can be arbitrarily long, and unless you cache the keys, you may be calling `oextract` a lot as well. Not only does this choice reduce the attractiveness of olist as a data representation, it adds very little, since the cases where olist makes sense pretty much always have a workable key. 

#### type olist

    [olist(T, U, oextract): elts]

where `elts: list(T)`.

    val olist = \typegen(T::entity, U::ordered, oextract:\(T)->U) {
        val OLT = olist(T, U, oextract)
        method index = \(oelt: U) -> integer {...}
        method insert = \mod(elt: T) -> list(T) {...}
        method insertIfUniq = \mod(elt: T) -> nothing {...}
        method count = \() -> integer {...}
    	method remove = \mod(elt: T) -> nothing {...}
        method $in = \(elt: T) -> boolean {...}
        method keyIn = \(oelt: U) -> boolean {...}
        method get = \(oelt: U) -> nilPossible(T) {...}
        method getOk = \(oelt: U) -> T {...}
        method before = \(elt: T) -> nilPossible(T) {...}
        method after = \(elt: T) -> nilPossible(T) {...}
        method rvalindex = \(inx: integer) -> T {...}
        method diff = \(other: OLT) -> OLT {...}
        method lithook4Set = \mod(elts: list(T)) -> OLT {...}
        method popb = \mod() -> nilPossible(T)
        method popf = \mod() -> nilPossible(T)
        method minelt = \() -> nilPossible(T)
        method maxelt = \() -> nilPossible(T)
    }

Method `index` provides direct access to what binary search returns. But use it carefully since any insert or remove can change the index.

Method `remove` removes the element, does nothing if element isn't present.

Method `$in` tells if the given elt is in the olist  (only correct if $< is a total order).

Method `keyIn` is like $in but only tests the key.

Method `get` returns the (first) elt whose oextract is oelt, or nil.

Method `getOk` is as get except you know the match will succeed so the return type isn't nilpossible.

Method `before` returns the greatest list elt <= T or nil if either list is empty or T is < the first elt of the list.

Method `after` returns the least list elt >= T or nil if either list is empty or T is > the last elt of the list.

Method `insertIfUniq` returns "unless(elt in self) insert(elt)", hand written because compiler doesn't do it well enough yet.

Method `rvalindex` can be helpful for types that use olists.

Method `diff` produces self - other, ie elts in self and not in other, in the manner of the set operation.

Method `lithook4Set` is defined to support stmapping from set to olist.

Method `popf` removes first element. (Note: nilPossible.)

Method `popb` removes last element. (Note: nilPossible.)

Method `minelt` returns first element. (Note: nilPossible.)

Method `maxelt` returns last element. (Note: nilPossible.)

#### type olist0

    val olist0 = \typegen(T0::ordered) {...}

`Olist0` simply instantiates `olist`:

    val olist0 = \typegen(T0::ordered) { olist(T0, T0, \(t: T0) {t}) }
