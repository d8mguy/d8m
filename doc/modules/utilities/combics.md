## Combics

Combics provides a selection of useful combinatorial algorithms. The idea is to add more definitions over time.

#### enumCombns:function

    val enumCombns : \(n, k, totcount: integer) -> list(list(integer))

`enumCombns` returns a maximum of `totcount` of the combinations of n choose k. For this to make sense, we need `k >= 1` and `k < n`; `enumCombns` panics unless these conditions are met. Each combination is a list(integer) of length k of integers between 0 and n (exclusive); in addition, each combination will have strictly increasing values and the list of combinations is ordered lexicographically.

#### randperm:function

    val randperm : \mod(list(T)) -> list(T) for any T::entity

Randomly permute a list of type `T`, using go package `math/rand` with implicit seed to generate the random numbers. It modifies and returns the list. Note that repeated calls to this function will not generate all permutations of the list without repetition. Note also that since it uses the PRNG state built into `math/rand`, using this may change the behavior of other program elements that also use that PRNG. If this is a problem, you can easily write
a version with its own state; the code is just 8 lines long. The algorithm used here is called the Fisher-Yates shuffle; as explained in the wikipedia, a version was published in 1938 by Fisher and Yates, and popularized in vol 2 of Knuth's The Art of Computer Programming.

#### streamIndices:type

    val streamIndices = extend tuple(cur, limits: list(integer)) with {
        method lithook : \mod(lims: list(integer))
        method eltcount : \() -> integer
        method nDimensions : \() -> integer
        method restart : \mod() -> nothing
        method done = \() -> boolean
        method current = \() -> list(integer)
        method next = \mod() -> list(integer)
    }

A stream type that accepts a list of lengths and generates all the indices up to each of those lengths, in
lexicographic order. Thus, given [1,2,3] it will generate [[0,0,0],[0,0,1],[0,0,2],[0,1,0],[0,1,1],[0,1,2]].
Note that `[streamIndices: ]` is immediately done, it generates nothing. This is the right way to handle the situation where you need to enumerate all the elements of a variable number of lists. It's sensible to think of the k
values that initialize a `streamIndices` as the extents of a k-dimensional cube. From this perspective, the stream generates all vertices of the cube in lexicographic order. A  `streamIndices` will panic if any initializing length
is <= 0 or `next` is called when `done`.

A `stream type` by convention has methods named `current`, `next`, and `done`; you call `next` until `done` becomes true. 