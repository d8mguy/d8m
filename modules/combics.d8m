// Combics provides a selection of useful combinatorial algorithms. This is intended to be an evolving module for quite some time.

export randperm, enumCombns, streamIndices

// Return a list(list(integer)) where each element is a length k list chosen from 0...n.
val enumCombns = \imp(n, k, totcount: integer) {
    if(k < 1 || k >= n) exit("bad args")
    var proto = (1..k).{index}
    var rslt: list(list(integer)) = [copy(proto)]
    val nkdiff = n - k
    var incrAt = k - 1
    // while condition is final exit criterion
    while(incrAt > 0 || proto[0] < n - k) {
        // loop through the options at current incrementing position
        while((incrAt < k - 1 && proto[incrAt] < proto[incrAt+1]+1) || proto[incrAt] < n - 1) {
            proto[incrAt] += 1
            rslt.pushb(copy(proto))
        }
        if(rslt.count > totcount) break
        // then reset incrAt and values in proto to set up next incrementing sequence
        while(incrAt > 0) {
            incrAt -= 1
            if(proto[incrAt] < nkdiff + incrAt) {
                var nval = proto[incrAt]+1
                each(h^incrAt...k) {
                    proto[h] = nval
                    nval += 1
                }
                rslt.pushb(copy(proto))
                incrAt = k - 1
                break
            }
        }
    }
    rslt
}

// A stream type that accepts a list of lengths and generates all the indices up to each of those lengths, in
// lexicographic order. Thus, given [1,2,3] it will generate [[0,0,0],[0,0,1],[0,0,2],[0,1,0],[0,1,1],[0,1,2]].
// Note that `[streamIndices: ]` is immediately done, it generates nothing. This is the right way to handle the
// situation where you need to enumerate all the elements of a variable number of lists. It's sensible to think of the k
// values that initialize a `streamIndices` as the extents of a k-dimensional cube. From this perspective, the
// stream generates all vertices of the cube in lexicographic order. A  `streamIndices` will panic if any initializing length
// is <= 0 or `next` is called when `done`.
//
// A `stream type` by convention has methods named `current`, `next`, and `done`; you call `next` until `done` becomes true.
val streamIndices = extend tuple(cur, limits: list(integer)) where {
    method lithook = \mod(lims: list(integer)) {
        if(any(lims.{this < 1})) exit("streamIndices requires positive limits")
        limits = lims.{this - 1}
        cur = zerolist(lims.count)
        unless(lims.count == 0) cur[lims.count - 1] = -1     // initialize so next returns all zeros first time
    }
    // utility method: number of elts produced by self from create to done
    method eltcount = \() { reduce(limits, \(a,b: integer) { a*(b+1)}, 1) }
    // utility method: number of "dimensions"
    method nDimensions = \() { cur.count }
    method restart = \mod() {
        cur = zerolist(limits.count)
        unless(limits.count == 0) cur[limits.count - 1] = -1
    }
    method done = \() { cur == limits }
    method current = \() { cur }
    method next = \mod() -> list(integer) {
        val nelts = limits.count
        each(i^1..nelts) {
            val ii = nelts - i; if(cur[ii] < limits[ii]) { cur[ii] += 1; each(j^(ii + 1)...nelts) cur[j] = 0; return cur }
        }
        exit("should not call terminated stream")
    }
}

// Randomly permute a list(T), using go package `math/rand` with implicit seed to generate the random numbers. It modifies and returns the list.
// Note that repeated calls to this function will not generate all its permutations without repetition.
// Note also that since it uses the PRNG state built into `math/rand`, using this may change
// the behavior of other program elements that also use that PRNG. If this is a problem, you can easily write
// a version with its own state; the code is just 8 lines long.
// The algorithm here is called the Fisher-Yates shuffle; as explained in the wikipedia, a version was
// published in 1938 by Fisher and Yates, and popularized in vol 2 of Knuth's TAOCP.

given(T::entity) {
    val randperm = \mod(lst: list(T), randSrc: \(integer)->integer) {
        var lim = count(lst)
        each(i^0...count(lst)-1) {
            val j = i+randSrc(lim)
            swap(lst, i, j)
            lim -= 1
        }
        lst
    }
}
