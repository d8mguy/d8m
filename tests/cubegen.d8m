
// longer than most tests, this is a fragment from generating projections on hypercubes for charting.
import "combics"
import "olist" melted
import go "strings"


val langNames = ["d8m", "go", "python", "ruby", "javascript", "c#", "java", "lua", "c++", "rust"];

// Parameter is used in a few ways. They live in the Metafile of each perfspec as descrns of the "normal" dimensions.
// ALso, server sends them after select to help define a chart. (And client sends back assignments to them.)
val Parameter = tuple(ident:string, explanation:string, elements:list(string))
val Output = tuple(ident: string, explanation: string, unit: string)

// Used in both client & server, this is the json decode of the meta.info file for a given perfspec working dir
val Metafile = tuple(ident, descrn: string, params: list(Parameter), outputs: list(Output),
	substns: list(Parameter), language: list(integer), spec: list(string), runtime: list(string), compiler: list(string))


val filedataT = tuple(rowcols: list(list(string)), filename: string)
val datafilesT = olist(filedataT, string, \(x:filedataT) { x.filename })
val vertexT = tuple(row, col: integer, data: filedataT)
// The list(cubeDim) in a canonicalCube always has the order [abscissa, multi, click]; the latter 2 can be omitted.
// The affectsRow slot can only be true for an abscissa.
val cubeDim = tuple(nElts:integer, offset: integer, affectsRow:boolean)
val canonicalCubeT = extend tuple(xvalues: list(string), absInx: integer, vertices:list(vertexT)) where {
    // generate the json form of a row for the chart embodied in self's vertices setup, given x info.
    method generateCols = \imp(xvalue: string, row0:integer) -> string {
        var rslt = [`{"x":"` + xvalue + `"`] + vertices.{ "\"y#{index}\":\"" + data.rowcols[row0+this.row][this.col] + `"` }
        strings.Join(rslt, ",") + "}"
    }
    method generateRows = \imp() {
        val rowValues = xvalues.{ generateCols(this, index) }
        "[" + strings.Join(rowValues, ",") + "]"
    }
}

// Given an allDimsIndices, return the basename of its filename
val allDims2Filename = \(adi: list(integer), nParams, nSubstns: integer) {
    val substns = cvt((0...nSubstns).{ val incr:byte = max(adi[nParams+this], 0) cvt('a'+ incr, byte) }, string)
    val bicds = strings.Join(adi[adi.count-5...adi.count-1].{ "#{this <= 0 ? 0 : this}"}, "_")
    "LM"+substns+bicds
}

// Given an index in the allDims of this metafile, access the right slot to get #elts
// Note that this doesn't handle repeats and assumes it won't be called.
val metafileIndex2Elts = \(mf: Metafile, index: integer) {
    val nParams = mf.params.count, nSubstns = mf.substns.count
    val nDims = nParams + nSubstns + 5
    if(index < nParams) mf.params[index].elements
    else if(index < nParams + nSubstns) mf.substns[index - nParams].elements
    else if(index == nDims - 5) mf.language.{ langNames[this] }
    else if(index == nDims - 4) mf.spec
    else if(index == nDims - 3) mf.runtime
    else mf.compiler
}

// Note that by construction (see doSelect) the order of allDims is (1) normal params; (2) substns; (3) BICDs in fixed order.
val selectedT = tuple(activePS: Metafile)
val stateAttribs = tuple(
    sel: selectedT,                  // info about the selected perfspec
    filedata: datafilesT            // this is the cache of filedata
)
val runstateT = extend stateAttribs where {

    // retrieve a filedataT from the cache; if not present, create and cache by reading in the file
    method getFiledata = \mod(filename: string, nrows, ncols: integer) -> filedataT {
        var cacheInx = filedata.index(filename)
        if(cacheInx < 0 || filedata[cacheInx].filename != filename) {
            // the real code will open and read a file; here, take the expectations & synth approp strgs
            var lines:list(list(string)) = []
            each(r^0...nrows) {
                val row = (0...ncols).{ "[#{r},#{this}]"+filename }
                lines.pushb(row)
            }
            filedata.insert([filedataT: lines, filename])
            cacheInx = filedata.index(filename)
        }
        filedata[cacheInx]
    }

    // Given the spec from client for a chart, generate the canonicalCube we use to read and format the data
    // Use (and fill in) the filedata cache as possible/needed.
    // For now, this can't handle dual axis, outs refers to the left axis, will need to add an arg (and more hair) later.
    method constructCube = \mod(allDimsIndices, outs: list(integer)) -> canonicalCubeT {
        var verts: list(vertexT) = []
        val absInx = always(x~allDimsIndices.index(-1), x!=nil)
        val multiInx = allDimsIndices.index(-3)
        val clickInx = allDimsIndices.index(-2)
        val nParams = sel.activePS.params.count
        val nSubstns = sel.activePS.substns.count
        var xstrgs: list(string) = []
        var absCount = 0
        // allDimsIndices.count = nParams+nSubstns+5
        // form a list of pairs of integers: elt count, index in allDims
        var filecubePairs: list(list(integer)) = []
        val absElts = metafileIndex2Elts(sel.activePS, absInx)
        var multiElts, clickElts: list(string)
        if(absInx >= nParams) filecubePairs.pushb([absElts.count, absInx])
        if(multiInx != nil && multiInx >= nParams) {
            multiElts = metafileIndex2Elts(sel.activePS, multiInx)
            filecubePairs.pushb([multiElts.count, multiInx])
        }
        if(clickInx != nil && clickInx >= nParams) {
            clickElts = metafileIndex2Elts(sel.activePS, clickInx)
            filecubePairs.pushb([clickElts.count, clickInx])
        }
        var fcstrm = [combics.streamIndices: filecubePairs.{this[0]}]
        // Now I'm ready to generate the filenames, collect the data with caching, assemble it into a proto cube
        var filenames: list(string) = []      // we'll build up this list by enumerating dimensions that involve diff files
        while(!fcstrm.done) {
            val nxfile:list(integer) = fcstrm.next    // nxfile.count == filecubePairs.count
            var adi0 = copy(allDimsIndices)
            each(x^xp([inx~nxfile, fcpair~filecubePairs])) {
                // fcpair[1] is the index in allDimsIndices of this dim; inx is the current value we want to put there
                adi0[x.fcpair[1]] = x.inx
            }
            filenames.pushb(allDims2Filename(adi0, nParams, nSubstns))
        }
        // At this point, filenames is a filled out cube wrt filename changes, of the full cube minus any repeated outputs
        // Use the caching scheme to fill in the desired canonicalCube from this.
        var vertices: list(vertexT) = []
        val nOuts = sel.activePS.outputs.count
        each(fnm^filenames) {
            each(outInx^outs) vertices.pushb([vertexT: 0, nParams + outInx, getFiledata(fnm, absElts.count, nParams + nOuts)])
        }
        [canonicalCubeT: absElts, 0, vertices]
    }
}

// mf1 has 1+1+5 params total and 4 outputs
val mf1 = [Metafile: "mf1", "mf1 desc", [[Parameter: "prm1", "", ["p11.0","p11.1","p11.2","p11.3"]]],
            [[Output: "out1.1", "", "u"], [Output: "out1.2", "", "u"], [Output: "out1.3", "", "u"], [Output: "out1.4", "", "u"]],
            [[Parameter: "subst11", "", ["s1v1", "s1v2", "s1v3", "s1v4"]]],
            [], ["spec11", "spec12", "spec13", "spec14", "spec15"], [], []]
// mf2 has 2+2+5 params total and 2 outputs
val mf2 = [Metafile: "mf2", "mf2 desc",
              [[Parameter: "prm2a", "", ["p21.0","p21.1","p21.2","p21.3", "p21.4"]], [Parameter: "prm2b", "", ["p22.0","p22.1","p22.2","p22.3"]]],
              [[Output: "out2.1", "", "u"], [Output: "out2.2", "", "u"]],
              [[Parameter: "subst21", "", ["s2v1", "s2v2", "s2v3"]], [Parameter: "subst22", "", ["s2bv1", "s2bv2", "s2bv3", "s2bv4"]]],
              [], ["spec21", "spec22", "spec23", "spec24", "spec25", "spec26", "spec27"], [], []]

val foo = \imp() {
    var rs1 = [runstateT: [selectedT: mf1], []]
    println(rs1.constructCube([-1, -2, 0, 1, 0, 0, 0], [2]).generateRows())
    println(rs1.constructCube([-1, 2, 0, -2, 0, 0, 0], [2]).generateRows())
    println(rs1.constructCube([-1, -3, 0, -2, 0, 0, 0], [2]).generateRows())
    var rs2 = [runstateT: [selectedT: mf2], []]
    println(rs2.constructCube([-1, 0, -2, 0, 0, 1, 0, 0, 0], [0,1]).generateRows())
    println(rs2.constructCube([-2, 0, -1, 0, 0, 1, 0, 0, 0], [0,1]).generateRows())
}
foo()
