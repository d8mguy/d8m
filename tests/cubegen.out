[{"x":"p11.0","y0":"[0,3]LMa0_1_0_0","y1":"[0,3]LMb0_1_0_0","y2":"[0,3]LMc0_1_0_0","y3":"[0,3]LMd0_1_0_0"},{"x":"p11.1","y0":"[1,3]LMa0_1_0_0","y1":"[1,3]LMb0_1_0_0","y2":"[1,3]LMc0_1_0_0","y3":"[1,3]LMd0_1_0_0"},{"x":"p11.2","y0":"[2,3]LMa0_1_0_0","y1":"[2,3]LMb0_1_0_0","y2":"[2,3]LMc0_1_0_0","y3":"[2,3]LMd0_1_0_0"},{"x":"p11.3","y0":"[3,3]LMa0_1_0_0","y1":"[3,3]LMb0_1_0_0","y2":"[3,3]LMc0_1_0_0","y3":"[3,3]LMd0_1_0_0"}]
[{"x":"p11.0","y0":"[0,3]LMc0_0_0_0","y1":"[0,3]LMc0_1_0_0","y2":"[0,3]LMc0_2_0_0","y3":"[0,3]LMc0_3_0_0","y4":"[0,3]LMc0_4_0_0"},{"x":"p11.1","y0":"[1,3]LMc0_0_0_0","y1":"[1,3]LMc0_1_0_0","y2":"[1,3]LMc0_2_0_0","y3":"[1,3]LMc0_3_0_0","y4":"[1,3]LMc0_4_0_0"},{"x":"p11.2","y0":"[2,3]LMc0_0_0_0","y1":"[2,3]LMc0_1_0_0","y2":"[2,3]LMc0_2_0_0","y3":"[2,3]LMc0_3_0_0","y4":"[2,3]LMc0_4_0_0"},{"x":"p11.3","y0":"[3,3]LMc0_0_0_0","y1":"[3,3]LMc0_1_0_0","y2":"[3,3]LMc0_2_0_0","y3":"[3,3]LMc0_3_0_0","y4":"[3,3]LMc0_4_0_0"}]
[{"x":"p11.0","y0":"[0,3]LMa0_0_0_0","y1":"[0,3]LMa0_1_0_0","y2":"[0,3]LMa0_2_0_0","y3":"[0,3]LMa0_3_0_0","y4":"[0,3]LMa0_4_0_0","y5":"[0,3]LMb0_0_0_0","y6":"[0,3]LMb0_1_0_0","y7":"[0,3]LMb0_2_0_0","y8":"[0,3]LMb0_3_0_0","y9":"[0,3]LMb0_4_0_0","y10":"[0,3]LMc0_0_0_0","y11":"[0,3]LMc0_1_0_0","y12":"[0,3]LMc0_2_0_0","y13":"[0,3]LMc0_3_0_0","y14":"[0,3]LMc0_4_0_0","y15":"[0,3]LMd0_0_0_0","y16":"[0,3]LMd0_1_0_0","y17":"[0,3]LMd0_2_0_0","y18":"[0,3]LMd0_3_0_0","y19":"[0,3]LMd0_4_0_0"},{"x":"p11.1","y0":"[1,3]LMa0_0_0_0","y1":"[1,3]LMa0_1_0_0","y2":"[1,3]LMa0_2_0_0","y3":"[1,3]LMa0_3_0_0","y4":"[1,3]LMa0_4_0_0","y5":"[1,3]LMb0_0_0_0","y6":"[1,3]LMb0_1_0_0","y7":"[1,3]LMb0_2_0_0","y8":"[1,3]LMb0_3_0_0","y9":"[1,3]LMb0_4_0_0","y10":"[1,3]LMc0_0_0_0","y11":"[1,3]LMc0_1_0_0","y12":"[1,3]LMc0_2_0_0","y13":"[1,3]LMc0_3_0_0","y14":"[1,3]LMc0_4_0_0","y15":"[1,3]LMd0_0_0_0","y16":"[1,3]LMd0_1_0_0","y17":"[1,3]LMd0_2_0_0","y18":"[1,3]LMd0_3_0_0","y19":"[1,3]LMd0_4_0_0"},{"x":"p11.2","y0":"[2,3]LMa0_0_0_0","y1":"[2,3]LMa0_1_0_0","y2":"[2,3]LMa0_2_0_0","y3":"[2,3]LMa0_3_0_0","y4":"[2,3]LMa0_4_0_0","y5":"[2,3]LMb0_0_0_0","y6":"[2,3]LMb0_1_0_0","y7":"[2,3]LMb0_2_0_0","y8":"[2,3]LMb0_3_0_0","y9":"[2,3]LMb0_4_0_0","y10":"[2,3]LMc0_0_0_0","y11":"[2,3]LMc0_1_0_0","y12":"[2,3]LMc0_2_0_0","y13":"[2,3]LMc0_3_0_0","y14":"[2,3]LMc0_4_0_0","y15":"[2,3]LMd0_0_0_0","y16":"[2,3]LMd0_1_0_0","y17":"[2,3]LMd0_2_0_0","y18":"[2,3]LMd0_3_0_0","y19":"[2,3]LMd0_4_0_0"},{"x":"p11.3","y0":"[3,3]LMa0_0_0_0","y1":"[3,3]LMa0_1_0_0","y2":"[3,3]LMa0_2_0_0","y3":"[3,3]LMa0_3_0_0","y4":"[3,3]LMa0_4_0_0","y5":"[3,3]LMb0_0_0_0","y6":"[3,3]LMb0_1_0_0","y7":"[3,3]LMb0_2_0_0","y8":"[3,3]LMb0_3_0_0","y9":"[3,3]LMb0_4_0_0","y10":"[3,3]LMc0_0_0_0","y11":"[3,3]LMc0_1_0_0","y12":"[3,3]LMc0_2_0_0","y13":"[3,3]LMc0_3_0_0","y14":"[3,3]LMc0_4_0_0","y15":"[3,3]LMd0_0_0_0","y16":"[3,3]LMd0_1_0_0","y17":"[3,3]LMd0_2_0_0","y18":"[3,3]LMd0_3_0_0","y19":"[3,3]LMd0_4_0_0"}]
[{"x":"p21.0","y0":"[0,2]LMaa0_1_0_0","y1":"[0,3]LMaa0_1_0_0","y2":"[0,2]LMba0_1_0_0","y3":"[0,3]LMba0_1_0_0","y4":"[0,2]LMca0_1_0_0","y5":"[0,3]LMca0_1_0_0"},{"x":"p21.1","y0":"[1,2]LMaa0_1_0_0","y1":"[1,3]LMaa0_1_0_0","y2":"[1,2]LMba0_1_0_0","y3":"[1,3]LMba0_1_0_0","y4":"[1,2]LMca0_1_0_0","y5":"[1,3]LMca0_1_0_0"},{"x":"p21.2","y0":"[2,2]LMaa0_1_0_0","y1":"[2,3]LMaa0_1_0_0","y2":"[2,2]LMba0_1_0_0","y3":"[2,3]LMba0_1_0_0","y4":"[2,2]LMca0_1_0_0","y5":"[2,3]LMca0_1_0_0"},{"x":"p21.3","y0":"[3,2]LMaa0_1_0_0","y1":"[3,3]LMaa0_1_0_0","y2":"[3,2]LMba0_1_0_0","y3":"[3,3]LMba0_1_0_0","y4":"[3,2]LMca0_1_0_0","y5":"[3,3]LMca0_1_0_0"},{"x":"p21.4","y0":"[4,2]LMaa0_1_0_0","y1":"[4,3]LMaa0_1_0_0","y2":"[4,2]LMba0_1_0_0","y3":"[4,3]LMba0_1_0_0","y4":"[4,2]LMca0_1_0_0","y5":"[4,3]LMca0_1_0_0"}]
[{"x":"s2v1","y0":"[0,2]LMaa0_1_0_0","y1":"[0,3]LMaa0_1_0_0","y2":"[0,2]LMba0_1_0_0","y3":"[0,3]LMba0_1_0_0","y4":"[0,2]LMca0_1_0_0","y5":"[0,3]LMca0_1_0_0"},{"x":"s2v2","y0":"[1,2]LMaa0_1_0_0","y1":"[1,3]LMaa0_1_0_0","y2":"[1,2]LMba0_1_0_0","y3":"[1,3]LMba0_1_0_0","y4":"[1,2]LMca0_1_0_0","y5":"[1,3]LMca0_1_0_0"},{"x":"s2v3","y0":"[2,2]LMaa0_1_0_0","y1":"[2,3]LMaa0_1_0_0","y2":"[2,2]LMba0_1_0_0","y3":"[2,3]LMba0_1_0_0","y4":"[2,2]LMca0_1_0_0","y5":"[2,3]LMca0_1_0_0"}]
